components/org.wso2.micro.integrator.extensions/org.wso2.micro.integrator.probes/src/main/java/org/wso2/micro/integrator/probes/ReadinessResource.java

```diff
@@ -0,0 +1,94 @@
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.micro.integrator.probes;
+
+import org.apache.axis2.AxisFault;
+import org.apache.axis2.Constants;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.synapse.MessageContext;
+import org.apache.synapse.commons.json.JsonUtil;
+import org.apache.synapse.core.axis2.Axis2MessageContext;
+import org.wso2.carbon.inbound.endpoint.internal.http.api.APIResource;
+import org.wso2.micro.integrator.core.services.CarbonServerConfigurationService;
+import org.wso2.micro.integrator.initializer.deployment.application.deployer.CAppDeploymentManager;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * API resource of the readiness probe API.
+ */
+public class ReadinessResource extends APIResource {
+
+    private static Log log = LogFactory.getLog(ReadinessResource.class);
+    public static final String NO_ENTITY_BODY = "NO_ENTITY_BODY";
+    public static final String HTTP_SC = "HTTP_SC";
+
+    /**
+     * Constructor for creating an API Resource.
+     *
+     * @param urlTemplate the url template of the Resource
+     */
+    public ReadinessResource(String urlTemplate) {
+        super(urlTemplate);
+    }
+
+    @Override
+    public Set<String> getMethods() {
+        Set<String> methods = new HashSet<>();
+        methods.add("GET");
+        return methods;
+    }
+
+
+    /**
+     * Health-check implementation.
+     * Returns 500 if faulty apps are detected.
+     */
+    @Override
+    public boolean invoke(MessageContext synCtx) {
+        org.apache.axis2.context.MessageContext axisCtx = ((Axis2MessageContext) synCtx).getAxis2MessageContext();
+        axisCtx.setProperty(Constants.Configuration.MESSAGE_TYPE, "application/json");
+        axisCtx.setProperty(Constants.Configuration.CONTENT_TYPE, "application/json");
+        // Sending a response body for a GET request
+        axisCtx.removeProperty(NO_ENTITY_BODY);
+        // appending MI server version number to the response
+        CarbonServerConfigurationService serverConfig = CarbonServerConfigurationService.getInstance();
+        String miVersion = serverConfig.getServerVersion();
+        String response = "{\"version\":\"" + miVersion + "\",";
+
+        ArrayList<String> faultyCapps = CAppDeploymentManager.getFaultyCapps();
+        if (faultyCapps.size() > 0) {
```

Comment:
Shall we add a configurable parameter here to return 200 even with faulty capss. i.e by default we will return 500 if fault but it can be configurable that we can return 200 if that system property is set to true. 
Then we can use this probe in Integration Studio as well to make sure that server is started ( currently we use Prometheus ) , in that case we can disable the system property and let server start with faulty capss.

Reply:
Have we tested the scenarios where we have multiple CAPPs and only one has failed? 
What if a particular CAPP takes a long time to deploy? Does API wait until that CAPP finish its deployment to return 200?

Reply:
If the purpose is only to check the server is started, tooling can just check the response payload != null.

Reply:
AFAIK the server plugin used in tooling uses this service and it needs a 200 status code.